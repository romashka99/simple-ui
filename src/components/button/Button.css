.button {
	display: inline-flex;
	justify-content: center;
	align-items: center;
	line-height: 1;
	height: 32px;
	white-space: nowrap;
	cursor: pointer;
	color: var(--button-text-color);
	text-align: center;
	box-sizing: border-box;
	outline: none;
	transition: 0.1s;
	font-weight: var(--button-font-weight);
	user-select: none;
	vertical-align: middle;
	-webkit-appearance: none;
	background-color: var(--button-bg-color);
	border: var(--border);
	border-color: var(--button-border-color);
	padding: 8px 15px;
	font-size: var(--font-size-base);
	border-radius: var(--border-radius-base);
}

.button {
	--button-font-weight: var(--font-weight-primary);
	--button-border-color: var(--border-color);
	--button-bg-color: var(--fill-color-blank);
	--button-text-color: var(--text-color-regular);
	--button-disabled-text-color: var(--disabled-text-color);
	--button-disabled-bg-color: var(--fill-color-blank);
	--button-disabled-border-color: var(--border-color-light);
	--button-divide-border-color: rgba(255, 255, 255, 0.5);
	--button-hover-text-color: var(--color-primary);
	--button-hover-bg-color: var(--color-primary-light-9);
	--button-hover-border-color: var(--color-primary-light-7);
	--button-active-text-color: var(--button-hover-text-color);
	--button-active-border-color: var(--color-primary);
	--button-active-bg-color: var(--button-hover-bg-color);
	--button-outline-color: var(--color-primary-light-5);
	--button-hover-link-text-color: var(--color-info);
	--button-active-color: var(--text-color-primary);
}

.button:hover,
.button:focus {
	color: var(--button-hover-text-color);
	border-color: var(--button-hover-border-color);
	background-color: var(--button-hover-bg-color);
	outline: none;
}

.button:active {
	color: var(--button-active-text-color);
	border-color: var(--button-active-border-color);
	background-color: var(--button-active-bg-color);
	outline: none;
}

.button.is-plain {
	--button-hover-text-color: var(--color-primary);
	--button-hover-bg-color: var(--fill-color-blank);
	--button-hover-border-color: var(--color-primary);
}

.button.is-round {
	border-radius: var(--border-radius-round);
	padding: 8px 15px;
}

.button.is-circle {
	border-radius: 50%;
	padding: 8px;
}

.button.is-link {
	border-color: transparent;
	color: var(--button-text-color);
	background: transparent;
	padding: 2px;
	height: auto;
}

.el-button.is-link:not([disabled]):active {
	color: var(--button-active-color);
	border-color: transparent;
	background-color: transparent;
}

.button:disabled,
.button[disabled] {
	color: var(--button-disabled-text-color);
	cursor: not-allowed;
	background-image: none;
	background-color: var(--button-disabled-bg-color);
	border-color: var(--button-disabled-border-color);
}

.button--primary {
	--button-text-color: var(--color-white);
	--button-bg-color: var(--color-primary);
	--button-border-color: var(--color-primary);
	--button-outline-color: var(--color-primary-light-5);
	--button-active-color: var(--color-primary-dark-2);
	--button-hover-text-color: var(--color-white);
	--button-hover-link-text-color: var(--color-primary-light-5);
	--button-hover-bg-color: var(--color-primary-light-3);
	--button-hover-border-color: var(--color-primary-light-3);
	--button-active-bg-color: var(--color-primary-dark-2);
	--button-active-border-color: var(--color-primary-dark-2);
	--button-disabled-text-color: var(--color-white);
	--button-disabled-bg-color: var(--color-primary-light-5);
	--button-disabled-border-color: var(--color-primary-light-5);
}

.button--primary.is-plain,
.button--primary.is-link {
	--button-text-color: var(--color-primary);
	--button-bg-color: var(--color-primary-light-9);
	--button-border-color: var(--color-primary-light-5);
	--button-hover-text-color: var(--color-white);
	--button-hover-bg-color: var(--color-primary);
	--button-hover-border-color: var(--color-primary);
	--button-active-text-color: var(--color-white);
}

.button--success {
	--button-text-color: var(--color-white);
	--button-bg-color: var(--color-success);
	--button-border-color: var(--color-success);
	--button-outline-color: var(--color-success-light-5);
	--button-active-color: var(--color-success-dark-2);
	--button-hover-text-color: var(--color-white);
	--button-hover-link-text-color: var(--color-success-light-5);
	--button-hover-bg-color: var(--color-success-light-3);
	--button-hover-border-color: var(--color-success-light-3);
	--button-active-bg-color: var(--color-success-dark-2);
	--button-active-border-color: var(--color-success-dark-2);
	--button-disabled-text-color: var(--color-white);
	--button-disabled-bg-color: var(--color-success-light-5);
	--button-disabled-border-color: var(--color-success-light-5);
}

.button--success.is-plain,
.button--success.is-link {
	--button-text-color: var(--color-success);
	--button-bg-color: var(--color-success-light-9);
	--button-border-color: var(--color-success-light-5);
	--button-hover-text-color: var(--color-white);
	--button-hover-bg-color: var(--color-success);
	--button-hover-border-color: var(--color-success);
	--button-active-text-color: var(--color-white);
}

.button--error {
	--button-text-color: var(--color-white);
	--button-bg-color: var(--color-danger);
	--button-border-color: var(--color-danger);
	--button-outline-color: var(--color-danger-light-5);
	--button-active-color: var(--color-danger-dark-2);
	--button-hover-text-color: var(--color-white);
	--button-hover-link-text-color: var(--color-danger-light-5);
	--button-hover-bg-color: var(--color-danger-light-3);
	--button-hover-border-color: var(--color-danger-light-3);
	--button-active-bg-color: var(--color-danger-dark-2);
	--button-active-border-color: var(--color-danger-dark-2);
	--button-disabled-text-color: var(--color-white);
	--button-disabled-bg-color: var(--color-danger-light-5);
	--button-disabled-border-color: var(--color-danger-light-5);
}

.button--error.is-plain,
.button--error.is-link {
	--button-text-color: var(--color-danger);
	--button-bg-color: var(--color-danger-light-9);
	--button-border-color: var(--color-danger-light-5);
	--button-hover-text-color: var(--color-white);
	--button-hover-bg-color: var(--color-danger);
	--button-hover-border-color: var(--color-danger);
	--button-active-text-color: var(--color-white);
}

.button--info {
	--button-text-color: var(--color-white);
	--button-bg-color: var(--color-info);
	--button-border-color: var(--color-info);
	--button-outline-color: var(--color-info-light-5);
	--button-active-color: var(--color-info-dark-2);
	--button-hover-text-color: var(--color-white);
	--button-hover-link-text-color: var(--color-info-light-5);
	--button-hover-bg-color: var(--color-info-light-3);
	--button-hover-border-color: var(--color-info-light-3);
	--button-active-bg-color: var(--color-info-dark-2);
	--button-active-border-color: var(--color-info-dark-2);
	--button-disabled-text-color: var(--color-white);
	--button-disabled-bg-color: var(--color-info-light-5);
	--button-disabled-border-color: var(--color-info-light-5);
}

.button--info.is-plain,
.button--info.is-link {
	--button-text-color: var(--color-info);
	--button-bg-color: var(--color-info-light-9);
	--button-border-color: var(--color-info-light-5);
	--button-hover-text-color: var(--color-white);
	--button-hover-bg-color: var(--color-info);
	--button-hover-border-color: var(--color-info);
	--button-active-text-color: var(--color-white);
}

.button--warning {
	--button-text-color: var(--color-white);
	--button-bg-color: var(--color-warning);
	--button-border-color: var(--color-warning);
	--button-outline-color: var(--color-warning-light-5);
	--button-active-color: var(--color-warning-dark-2);
	--button-hover-text-color: var(--color-white);
	--button-hover-link-text-color: var(--color-warning-light-5);
	--button-hover-bg-color: var(--color-warning-light-3);
	--button-hover-border-color: var(--color-warning-light-3);
	--button-active-bg-color: var(--color-warning-dark-2);
	--button-active-border-color: var(--color-warning-dark-2);
	--button-disabled-text-color: var(--color-white);
	--button-disabled-bg-color: var(--color-warning-light-5);
	--button-disabled-border-color: var(--color-warning-light-5);
}

.button--warning.is-plain,
.button--warning.is-link {
	--button-text-color: var(--color-warning);
	--button-bg-color: var(--color-warning-light-9);
	--button-border-color: var(--color-warning-light-5);
	--button-hover-text-color: var(--color-white);
	--button-hover-bg-color: var(--color-warning);
	--button-hover-border-color: var(--color-warning);
	--button-active-text-color: var(--color-white);
}

.button--small {
	--button-size: 24px;
	height: var(--button-size);
	padding: 5px 11px;
	font-size: 12px;
	border-radius: calc(var(--border-radius-base) - 1px);
}

.button--large {
	--button-size: 40px;
	height: var(--button-size);
	padding: 12px 19px;
	font-size: var(--font-size-base);
	border-radius: var(--border-radius-base);
}
