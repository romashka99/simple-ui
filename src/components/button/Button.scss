@import '../../styles/colors.scss';
@import '../../styles/main.scss';
@import '../../styles/style.scss';

.button {
	display: flex;
	justify-content: center;
	align-items: center;
	height: 40px;
	padding: 10px 30px;

	cursor: pointer;

	background-color: $color-grey;
	color: $color-white;
	border: $border;
	border-radius: $border-radius;

	font-weight: 700;
	font-size: 14px;
	line-height: 20px;

	&.is-round {
		border-radius: 100px;
	}

	&.is-plain {
		background-color: $color-white;
		color: $color-grey;
	}

	&:disabled {
		cursor: not-allowed;
		background-color: $color-mid-light-grey;
		border-color: $color-mid-light-grey;
	}

	&:not(:disabled) {
		&:not(.is-plain) {
			&:hover,
			&:focus {
				background-color: $color-mid-grey;
				border-color: $color-mid-grey;
			}

			&:active {
				background-color: $color-dark-grey;
				border-color: $color-dark-grey;
			}
		}

		&.is-plain {
			&:hover,
			&:focus {
				color: $color-mid-grey;
				border-color: $color-mid-grey;
			}

			&:active {
				color: $color-dark-grey;
				border-color: $color-dark-grey;
			}
		}
	}
}
