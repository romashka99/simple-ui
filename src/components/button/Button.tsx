import React, { ReactNode } from 'react';
import { Color, Size } from '../../types/types';
import './Button.scss';

interface IButtonProps {
	className?: string;
	color?: Color;
	size?: Size;
	plain?: boolean;
	round?: boolean;
	circle?: boolean;
	link?: boolean;
	disabled?: boolean;
	children?: ReactNode;
	onClick?: (event: React.MouseEvent<HTMLButtonElement>) => void;
}

export default function Button({
	className,
	color = 'default',
	size = 'medium',
	plain,
	round,
	circle,
	link,
	disabled,
	onClick,
	children,
}: IButtonProps) {
	const classes = `button button--${color} button--${size}${
		plain ? ' is-plain' : ''
	}${round ? ' is-round' : ''}${circle ? ' is-circle' : ''}${
		link ? ' is-link' : ''
	}${className ? ' ' + className : ''}`;

	return (
		<button className={classes} onClick={onClick} disabled={disabled}>
			{children}
		</button>
	);
}
