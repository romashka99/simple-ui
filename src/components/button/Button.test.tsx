import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import React from 'react';
import Button from './Button';

describe('Test Button', () => {
	describe('default', () => {
		test('without attr', () => {
			render(<Button>Click</Button>);
			const button = screen.getByRole('button');
			expect(button.classList.length).toBe(3);
			expect(button.classList.contains('button')).toBe(true);
			expect(button.classList.contains('button--default')).toBe(true);
			expect(button.classList.contains('button--medium')).toBe(true);
			expect(button.getAttribute('disabled')).toBeNull();
			expect(button).toMatchSnapshot();
		});

		test('plain', () => {
			render(<Button plain>Click</Button>);
			const button = screen.getByRole('button');
			expect(button.classList.length).toBe(4);
			expect(button.classList.contains('button')).toBe(true);
			expect(button.classList.contains('button--default')).toBe(true);
			expect(button.classList.contains('button--medium')).toBe(true);
			expect(button.classList.contains('is-plain')).toBe(true);
			expect(button.getAttribute('disabled')).toBeNull();
			expect(button).toMatchSnapshot();
		});

		test('round', () => {
			render(<Button round>Click</Button>);
			const button = screen.getByRole('button');
			expect(button.classList.length).toBe(4);
			expect(button.classList.contains('button')).toBe(true);
			expect(button.classList.contains('button--default')).toBe(true);
			expect(button.classList.contains('button--medium')).toBe(true);
			expect(button.classList.contains('is-round')).toBe(true);
			expect(button.getAttribute('disabled')).toBeNull();
			expect(button).toMatchSnapshot();
		});

		test('disabled', () => {
			render(<Button disabled>Click</Button>);
			const button = screen.getByRole('button');
			expect(button.classList.length).toBe(3);
			expect(button.classList.contains('button')).toBe(true);
			expect(button.classList.contains('button--default')).toBe(true);
			expect(button.classList.contains('button--medium')).toBe(true);
			expect(button.getAttribute('disabled')).not.toBeNull();
			expect(button).toMatchSnapshot();
		});

		test('link', () => {
			render(<Button link>Click</Button>);
			const button = screen.getByRole('button');
			expect(button.classList.length).toBe(4);
			expect(button.classList.contains('button')).toBe(true);
			expect(button.classList.contains('button--default')).toBe(true);
			expect(button.classList.contains('button--medium')).toBe(true);
			expect(button.classList.contains('is-link')).toBe(true);
			expect(button.getAttribute('disabled')).toBeNull();
			expect(button).toMatchSnapshot();
		});

		test('small', () => {
			render(<Button scale='small'>Click</Button>);
			const button = screen.getByRole('button');
			expect(button.classList.length).toBe(3);
			expect(button.classList.contains('button')).toBe(true);
			expect(button.classList.contains('button--default')).toBe(true);
			expect(button.classList.contains('button--small')).toBe(true);
			expect(button.getAttribute('disabled')).toBeNull();
			expect(button).toMatchSnapshot();
		});

		test('large', () => {
			render(<Button scale='large'>Click</Button>);
			const button = screen.getByRole('button');
			expect(button.classList.length).toBe(3);
			expect(button.classList.contains('button')).toBe(true);
			expect(button.classList.contains('button--default')).toBe(true);
			expect(button.classList.contains('button--large')).toBe(true);
			expect(button.getAttribute('disabled')).toBeNull();
			expect(button).toMatchSnapshot();
		});

		test('custom className', () => {
			render(<Button className='test-class'>Click</Button>);
			const button = screen.getByRole('button');
			expect(button.classList.length).toBe(4);
			expect(button.classList.contains('button')).toBe(true);
			expect(button.classList.contains('button--default')).toBe(true);
			expect(button.classList.contains('button--medium')).toBe(true);
			expect(button.classList.contains('test-class')).toBe(true);
			expect(button.getAttribute('disabled')).toBeNull();
			expect(button).toMatchSnapshot();
		});

		test('onClick', () => {
			let value = 0;
			render(
				<Button
					onClick={() => {
						value = 1;
					}}
				>
					Click
				</Button>
			);
			const button = screen.getByRole('button');
			expect(button.classList.length).toBe(3);
			expect(button.classList.contains('button')).toBe(true);
			expect(button.classList.contains('button--default')).toBe(true);
			expect(button.classList.contains('button--medium')).toBe(true);
			expect(button.getAttribute('disabled')).toBeNull();
			expect(value).toBe(0);
			userEvent.click(button);
			expect(value).toBe(1);
			expect(button).toMatchSnapshot();
		});
	});

	describe('color', () => {
		test('primary', () => {
			render(<Button color='primary'>Click</Button>);
			const button = screen.getByRole('button');
			expect(button.classList.length).toBe(3);
			expect(button.classList.contains('button')).toBe(true);
			expect(button.classList.contains('button--primary')).toBe(true);
			expect(button.classList.contains('button--medium')).toBe(true);
			expect(button.getAttribute('disabled')).toBeNull();
			expect(button).toMatchSnapshot();
		});

		test('success', () => {
			render(<Button color='success'>Click</Button>);
			const button = screen.getByRole('button');
			expect(button.classList.length).toBe(3);
			expect(button.classList.contains('button')).toBe(true);
			expect(button.classList.contains('button--success')).toBe(true);
			expect(button.classList.contains('button--medium')).toBe(true);
			expect(button.getAttribute('disabled')).toBeNull();
			expect(button).toMatchSnapshot();
		});

		test('error', () => {
			render(<Button color='error'>Click</Button>);
			const button = screen.getByRole('button');
			expect(button.classList.length).toBe(3);
			expect(button.classList.contains('button')).toBe(true);
			expect(button.classList.contains('button--error')).toBe(true);
			expect(button.classList.contains('button--medium')).toBe(true);
			expect(button.getAttribute('disabled')).toBeNull();
			expect(button).toMatchSnapshot();
		});

		test('info', () => {
			render(<Button color='info'>Click</Button>);
			const button = screen.getByRole('button');
			expect(button.classList.length).toBe(3);
			expect(button.classList.contains('button')).toBe(true);
			expect(button.classList.contains('button--info')).toBe(true);
			expect(button.classList.contains('button--medium')).toBe(true);
			expect(button.getAttribute('disabled')).toBeNull();
			expect(button).toMatchSnapshot();
		});

		test('warning', () => {
			render(<Button color='warning'>Click</Button>);
			const button = screen.getByRole('button');
			expect(button.classList.length).toBe(3);
			expect(button.classList.contains('button')).toBe(true);
			expect(button.classList.contains('button--warning')).toBe(true);
			expect(button.classList.contains('button--medium')).toBe(true);
			expect(button.getAttribute('disabled')).toBeNull();
			expect(button).toMatchSnapshot();
		});
	});
});
