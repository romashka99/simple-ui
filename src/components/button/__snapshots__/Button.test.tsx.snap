// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Button default custom className 1`] = `
<button
  class="button  test-class"
  data-testid="button"
>
  Click
</button>
`;

exports[`Test Button default disabled 1`] = `
<button
  class="button "
  data-testid="button"
  disabled=""
>
  Click
</button>
`;

exports[`Test Button default onClick 1`] = `
<button
  class="button "
  data-testid="button"
>
  Click
</button>
`;

exports[`Test Button default onClick with disabled 1`] = `
<button
  class="button "
  data-testid="button"
  disabled=""
>
  Click
</button>
`;

exports[`Test Button default plain 1`] = `
<button
  class="button  is-plain"
  data-testid="button"
>
  Click
</button>
`;

exports[`Test Button default round 1`] = `
<button
  class="button  is-round"
  data-testid="button"
>
  Click
</button>
`;

exports[`Test Button default without attr 1`] = `
<button
  class="button "
  data-testid="button"
>
  Click
</button>
`;
